returnn/tf/util/data.py

```diff
@@ -914,6 +914,11 @@ def declare_same_as(self, other):
       self_same_as.declare_same_as(other_same_base)
       if (self.dyn_size_ext is None or not self._validate_in_current_graph()) and self_same_as.dyn_size_ext:
         self.dyn_size_ext = self_same_as.get_dyn_size_ext_for_batch_ctx(self.batch, self.control_flow_ctx)
+    self_derived_bases = set(self.get_derived_bases_list())
+    other_derived_bases = set(other.get_derived_bases_list())
```

Comment:
There should be an assert now like:
```python
assert self_derived_bases != other_derived_bases
```

Reply:
I don't understand. Why can it be equal? Then we should have `self == other` or not?

Reply:
I don't know exactly. I will look into it a post a test case with explanation. However it is probably the case that `self` has an ancestor that is equal to `other` but not the same object.  Similarly `other` has an ancestor that is equal to `self` but is not the same object. Then the two sets are equal but not containing the same objects. Thus we are running into the assertion.

Reply:
But equality (`==`, `!=`) of a set does check the equality of the set, i.e. equality of the objects (`==`), not identity (`is`).

Reply:
Right, that was what I was saying. The dimensions can be equal (`==`) without being identical (`is`). If the assertion would check for identity it would not fail (I believe).

Reply:
But `assert self_derived_bases != other_derived_bases` uses `!=` (`==`) not `is`.

Reply:
Yes. Because it uses `==` and not `is` the assertion fails. 
> However it is probably the case that self has an ancestor that is equal to other but not the same object. Similarly other has an ancestor that is equal to self but is not the same object. Then the two sets are equal but not containing the same objects. Thus we are running into the assertion.

Here I tried to describe a case where the equality (`==`) of the two sets is given but the sets are not identical (`is`). So even if `self` and `other` are not equal (`!=`) the two sets can still be equal (`==`) in the end (because both have ancestors that are equal to the other). But the sets are not identical. 

Or am I getting something wrong here?

Reply:
`self_derived_bases` also includes `self`, and `other_derived_bases` also includes `other`.

Reply:
> Why do you talk about identity (is) at all?

If the dimension tags are equal (as the assertion suggests) but we are not having the recursion problem that we are trying to solve, then the two sets must contain objects that are not identical. At least that's how I make sense of the situation. 

The problem can be reproduced like this:
```python
from returnn.tf.network import TFNetwork
from returnn.config import Config

import numpy as np
batch_size = 8
time_size = 20
feat_size = 10
shift_amount = 5  # right-shift of 5 elements
config = Config()
config.update({
  "num_outputs": feat_size,
  "num_inputs": feat_size,
  "network": {
    "output": {"class": "shift_axis", "from": ["data"], "amount": shift_amount, "pad": True, "axis": "T"}
  }
})
network = TFNetwork(config=config, train_flag=True)
network.construct_from_dict(config.typed_dict["network"])
```

The problematic line is https://github.com/rwth-i6/returnn/blob/ae0cc8c4f802013d69a86782b7dce30fb6db16f3/returnn/tf/util/data.py#L914

which is called after our check `self==other` but which might make `self` and `other` equal so that our two sets can then be equal again. We just have to do the subset check earlier in the code to avoid this problem I think.

Reply:
Why did you mark this as resolved? Is it resolved?

Reply:
Yes, I added the assertion and it seems to work now. Do you disagree?

Reply:
No I think it's fine.

Overall I'm not too happy with the whole function. It's way too complicated. But not sure how to simplify it. But anyway, this is not about this PR now.